### PR TITLE
Add support when writing to locked buckets by handling retentionPolicyNotMet error

### PR DESCRIPTION
**Pull request by ahmedabu98**
There is an edge case when Beam issues multiple rewrites to a GCS bucket. A bundle that rewrites files can succeed, but the fact of its success may not persist in Beam (e.g. autoscaling workers, worker restart). On bundle retry, the rewrite is issued again with the same destination. If this happens when writing to a bucket with a retention policy, we will run into the retentionPolicyNotMet error.

PR adds support for this case by handling the retentionPolicyNotMet error as such:
- If source and destination files have the same checksum, assume they are identical and skip the rewrite
- otherwise, throw an early error

More information in this design document: https://docs.google.com/document/d/11kXzI90KmAyknszSFmtfPcL_GWaVzpt8MojQfifZOoM/edit#heading=h.f8k9hty17wf5

### CODE REVIEW

**Comment by ahmedabu98** [comment]
R: @johnjcasey

**Comment by johnjcasey** [comment]
LGTM, very nice work

**Comment by johnjcasey** [comment]
run java precommit

**Comment by ahmedabu98** [comment]
Run Java PostCommit

**Comment by ahmedabu98** [comment]
Run PostCommit_Java_DataflowV2

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java [hunk 1]
```diff
@@ -897,6 +898,35 @@ public void onFailure(GoogleJsonError e, HttpHeaders responseHeaders) throws IOE
         } else {
           throw new FileNotFoundException(from.toString());
         }
+      } else if (e.getCode() == 403
+          && e.getErrors().size() == 1
+          && e.getErrors().get(0).getReason().equals("retentionPolicyNotMet")) {
+        List<StorageObjectOrIOException> srcAndDestObjects = getObjects(Arrays.asList(from, to));
+        if (srcAndDestObjects
+            .get(0)
```
Comment (chamikaramj) [inline sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java:906]: Are both the source and the destination guaranteed to exist at this point ?

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java [hunk 2]
```diff
@@ -897,6 +898,35 @@ public void onFailure(GoogleJsonError e, HttpHeaders responseHeaders) throws IOE
...
+            .storageObject()
+            .getMd5Hash()
+            .equals(srcAndDestObjects.get(1).storageObject().getMd5Hash())) {
```
Comment (chamikaramj) [inline sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java:909]: Please make sure that this does not pass trivially due to default values matching.

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java (hunk 1 above)
Comment (ahmedabu98) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java:906]: We get `retentionPolicyNotMet` error when we prematurely (1) try to overwrite an existing file or (2) try to delete an existing file: https://cloud.google.com/storage/docs/bucket-lock#retention-policy

Deletions are handled in a separate `onFailure()` method [here](https://github.com/apache/beam/blob/209daedff3e60f86188bc9e2d3e7e331e283373b/sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java#L815-L843), so at this point I believe we can assume both source and destination exist.

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java (hunk 2 above)
Comment (ahmedabu98) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java]: yeah SGTM!

**Comment by johnjcasey** [comment]
run java precommit

**Comment by johnjcasey** [comment]
run java postcommit

**Comment by chamikaramj** [comment]
Run PostCommit_Java_DataflowV2

**Comment by chamikaramj** [comment]
Run PostCommit_Java_DataflowV2

**Comment by ahmedabu98** [comment]
Run PostCommit_Java_DataflowV2

**Comment by chamikaramj** [comment]
Run Java_Examples_Dataflow PreCommit

**Pull request merged**